PIM SM HLD

vidyac86 opened the pull request:
This pull-request has the design document for PIM-ASM feature. The document has details about
1) Requirements (Functional/Management)
2) Use cases 
3) Description about the various Interfacing components
4) KAT timer support from different modules for PIM-ASM
5) Broadcom requirements 

Please check the design document and share your comments.

vidyac86 committed:
PIM SM HLD

ashok-krishnegowda commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+   
+   /** IPMC entry aged */
+   
+    SAI_IPMC_EVENT_AGED,
```
change this to -> SAI_IPMC_EVENT_AGE

ravisekhar-akkineni replied:
Modified

ashok-krishnegowda commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+|  mfpm_route_info_t | route_info | Info to get Route status from MFPM client |
+
+### 3.2.1 **SAI specific data structure and callback changes**
+
```
Add about new SAI attributes - SAI_SWITCH_ATTR_IPMC_EVENT_NOTIFY & SAI_SWITCH_ATTR_IPMC_AGING_TIME

More description below,
The IPMC registers a callback named sai_ipmc_event_notification_fn through the switch attribute SAI_SWITCH_ATTR_IPMC_EVENT_NOTIFY.
The SONiC then create/sets the IPMC entry with the optional attribute SAI_SWITCH_ATTR_IPMC_AGING_TIME.
ipmc_entry_attr[0].id = SAI_SWITCH_ATTR_IPMC_AGING_TIME;
ipmc_entry_attr[0].value.u32 = aging_time; // in seconds
If the attribute is not added then the IPMC entry does not age out. If the aging time is specified as 0, the IPMC entry does not age out. The aging time configured can be queried using a GET request of this attribute.
Once the IPMC entry ages out, SAI notifies SONiC using sai_ipmc_event_notification_fn callback. This callback provides events of type sai_ipmc_event_notification_data_t which has information about the specific IPMC entries that got aged out. The SONiC uses the IPMC entry information to delete these IPMC entries.

Gokulnath-Raja replied:
@ashok-krishnegowda based on our earlier discussion SAI_SWITCH_ATTR_IPMC_AGING_TIME is global config. Please refer below discussion point.
SAI API sai_set_switch_attribute() will be used to set the IPMC aging time with the attribute mentioned below,
sai_get_switch_attribute() to get the configured value.
 
Below SAI switch attribute will be added for the IPMC age time setting, 
 
/**
     * @brief Dynamic IPMC entry aging time in seconds
     *
     * Zero means aging is disabled.
     *
     * @type sai_uint32_t
     * @flags CREATE_AND_SET
     * @default 0
     */
    SAI_SWITCH_ATTR_IPMC_AGING_TIME,

ashok-krishnegowda commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+   1.1.(a) When the LHR receives an IGMP Report, it is passed to the Linux kernel.  Kernel delivers the IGMP Report payload to the Pimd process.  Pimd creates a IGMP group membership entry and checks if the RP address is reachable.  If reachable, Pimd creates a (\*,G) entry with the RPF interface as its IIF and updates the (\*,G) entry's OIF with the interface on which the IGMP Report is received ...
+   Data packets are needed by Pimd to do SPT switchover in case of LHR, To achieve this, the OIF list is updated with the pimreg interface. <br> <br>
+
+    1.1.(b) When the RP or an intermediate PIM router receives a PIM (\*,G) Join message, the message is passed to the Linux kernel and kernel delivers the PIM Join message and its associated payload to Pimd process.  PIM queries Zebra for RPF validation and then creates a (\*,G) entry with the RPF interface as its IIF. In case of RP, the IIF interface is the interface on which the RP address is  ...
```
In case of RP, the IIF interface is the interface on which the RP address is configured
   Ashok>> If loopback interface address is configured as RP then how does OA interprets this loopback in entry IIF ?

vidyac86 replied:
As loopback interface does not have a valid RPF, OA will not push this (*,G) entry to SAI. Updated this section as well as the usecase section with this details

ashok-krishnegowda commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+    
+11. The ipmcFpmSyncd process will be registered for the event notification for the multicast routes in the state DB. Once the DONT_AGE/AGE status is updated in the DB, ipmcFpmSyncd will receive this event and it will update it's local DB with the age and dont_age counter values.This DB would be updated as and when there is a change in the status of the routes in the Hardware. <br><br>
+    
+12. FRR which runs the KAT timer for each of the routes will query the ipmcFpmsyncd for the routes status. IpmcFpmSyncd will do the route status calculation based on the AGE/DONT_AGE counters and update the status to Pimd.<br><br>
```
Pls add more details on FRR to ipmcFpmsyncd query channel for the route status.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+| IIF       | Incoming interface                                                |
+| IGMP      | Internet Group Management Protocol                                |
+| LHR       | Last Hop Router (Router directly connected to the Host/Receiver   |
+| MLD       | Multicast Discovery Protocol                                      |
```
Correction - Multicast **Listener** Discovery

vidyac86 replied:
Corrected

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+    
+12. FRR which runs the KAT timer for each of the routes will query the ipmcFpmsyncd for the routes status. IpmcFpmSyncd will do the route status calculation based on the AGE/DONT_AGE counters and update the status to Pimd.<br><br>
+
+13. Pimd will process this route status and if the route status remains as aged for (2*KAT) value then the routes are purged in the TIB and deleted from Kernel and Hardware too.<br><br>
```
Why is this aging 2* KAT? is this from RFC or FRR recommends/behaves like this? Please explain more on this

vidyac86 replied:
No,it is not a RFC recommendation. This would be the worst case value for a route to be in Software if toggling is seen. Updated the section with the details.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+   
+   __Figure 3: PIM ASM Architecture in SONiC__
+
+The interactions between FRR PIM, Linux kernel, ipmcFpmSyncd and ipmcOrch for updating a multicast (*,G) and (S,G) route entry from PIM to the hardware are explained below:
```
PIM has (*,G), (S,G) and (S,G) with RPT(known as SG Rpt) . how is this handled? kernel does not have this? So from PIM SG and SG Rpt is treated as separate 2 SG entries if IIF are different? what will happen if IIF is same, Please explain more on this how SGRpt is handled

plsaranya replied:
With respect to FRR PIM, there would not be 2 different SG entries.  The SG entries would be with IIF as RPF(S) in shortest path tree and SG entry would be with IIF as RPF(RP) (SGRpt) in RP tree.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+    1.1.(b) When the RP or an intermediate PIM router receives a PIM (\*,G) Join message, the message is passed to the Linux kernel and kernel delivers the PIM Join message and its associated payload to Pimd process.  PIM queries Zebra for RPF validation and then creates a (\*,G) entry with the RPF interface as its IIF. In case of RP, the IIF interface is the interface on which the RP address is  ...
+    
+2. After PIM creates the (\*,G) entry with the required information, PIM updates Linux kernel using a socket and kernel creates a mroute entry. <br> <br>
```
at RP (*,G) entry will not have IIF , PIM protocol will have entry with IIF NULL, is that entry gets updated/sent to kernel? or only SG will be updated to kernel, please add more detail on this case at RP

vidyac86 replied:
FRR,doesn't have iif as NULL in RP. It has the interface on which the RP address is configured. This is passed to the Kernel and to ipmcfpmsyncd. But OA, is not installing this route as it doesn't have a valid RPF. Added details about this.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+13. Pimd will process this route status and if the route status remains as aged for (2*KAT) value then the routes are purged in the TIB and deleted from Kernel and Hardware too.<br><br>
+
+### 3.1.1 ipmcFpmSyncd 
+
```
Does kernel knows about "pimreg" entry how does kernel treats "pimreg" intf.

plsaranya replied:
yes, pimreg interface is a kernel interface which would be set with VIFF_REGISTER. So only based on pimreg interface the IGMP_WHOLEPKT kernel upcalls would be triggered

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+1. PIM ASM creates (S,G) entry once the data traffic starts to flow. <br> <br>
+  1.1 When FHR receives the unknown multicast data from a source it traps and delivers the multicast data to the Linux kernel. Kernel delivers this packet to the Pimd process. On reception of data, Pimd starts the registration process. 
+As FHR is not aware of the interested receivers,it encapsulates the data packet in a unicast PIM register message and forwards it to RP address. <br> <br> 
+  1.2  RP's kernel receives the PIM register packet. It decapsulates the data and forwards the data to the downstream routers and it delivers the register packet to Pimd. Pimd processes the register message and it creates (S,G) join by inheriting the OIF from the (\*,G) route along with cpu port and IIF as the incoming interface of the data and  forwards it towards the FHR.
```
Please confirm kernel forwards the Reg decap packet or pimd.  if there is detailed design document or any document about FRR behavior, its good idea to have reference in this document.

plsaranya replied:
FRR expects kernel to decapsulate and forward the packet. For now, kernel is not forwarding the packet because of a bug. That issue has been fixed in kernel and patch would be applied to achieve this

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+#### 3.1.2.2 IIF expansion in OIF List
+
+IIF member ports to be included in IPMC group member on following cases,
+1. L3 multicast case, local receiver on the IIF vlan then add IIF vlan in OIF list
```
Not clear why this is needed, SAI has SAI_IPMC_ENTRY_ATTR_RPF_GROUP_ID object, the IIF would be matched to this and the ports from IIF(in case of VLAN) would go to IPMC l2 bitmap right? which module this happens, hope its not form common code.  if its inside NPU specific code its fine, otherwise it would be fine. or is it current SAI module does not have way to specify this?

Gokulnath-Raja replied:
@mhampasagar  following are the discussion points with @ashok-krishnegowda on this requirement, and concluded on Approach1.
**Approach1:** 
IIF-VLAN will be always added as part of OIF-list by Orchagent
1.	If snooping is disabled, Orchagent will not fill L2MC-group-id for that IIF-VLAN.
          a.	SAI fills all VLAN-members today for these OIFs, but it fills it in L3-port-bitmap which modifies VLAN-tag.
          b.	In IIF-VLAN being OIF case, SAI should add VLAN-members as L2-port-bitmap. SAI should identify IIF-VLAN being OIF-VLAN by comparing RPF-group.
2.	If snooping is enabled, Orchagent will fill L2MC-group-id for that IIF-VLAN as like any other OIF-VLAN with snooped ports
**Approach2:**
To use RPF group member  which is set to IIF. Here if snooping is disabled, RPF member will be the same data as filled today,   SAI need to include IIF vlan member ports to multicast group in L2 port bitmap.  If in IIF vlan, snooping is enabled, SAI_IPMC_GROUP_MEMBER_ATTR_EXTENSIONS_L2MC_OUTPUT_ID should be included in RPF Group member, so that orchagent can fill it and communicate to SAI.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+IIF member ports to be included in IPMC group member on following cases,
+1. L3 multicast case, local receiver on the IIF vlan then add IIF vlan in OIF list
+2. Shared case (L3 and Snooping enabled), local receiver on the IIF vlan then add IIF vlan member ports into OIF list
```
With respect to L3 multicast and Snooping enabled case, more explanation might be needed as both can have same entry and what happens if snooping on specific VLAN enables/disables and without snooping enabled and just with PIM. if its captured in some other document , please share it.

Gokulnath-Raja replied:
In Orchagent, there are two cases handled (1 Only PIM enabled, 2 Both PIM and Snooping enabled).... Both case for if IIF is VLAN, VLAN is updated in OIF list.
ipmcorchagent, upon receiving the update adds IIF VLAN to OIF list(class MRouteEntry and member list m_oifs updated with IIF vlan)
a) calls IPMC group creation incase of L3 VLAN
b) call L2MC group creation incase of snooping enabled on the VLAN

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+   DONTAGE notifications.
+   IPMC specific notification will be identified with the enum "SAI_SWITCH_NOTIFICATION_NAME_IPMC_EVENT" which will be added 
+   newly in SAI layer. 
+3. On receiving the IPMC entry/entries AGE or DONTAGE data, required validations will be done after 
```
Hope both SAI and syncd handled bulk update for KAT update

ravisekhar-akkineni replied:
My understand is that whether SAI and SYNCD are designed to handle multiple IPMC entries aliveness data notifications? 

Yes, both SAI and SYNCD are designed to handle IPMC entries bulk data similar to FDB entries handling.  As part of callback, SAI provides the notification count and data array.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+ssm-ranges                : Configure Source-Specific-Multicast group range using IP Prefix-list
+ecmp-enabled              : To enable PIM ECMP (true/false)
+ecmp-rebalance-enabled    : To enable PIM ECMP rebalance (true/false). It can be enabled/disabled only when ECMP is enabled
+spt-threshold             : To disable or re-enable the SPT switchover. By default, SPT switchover is enabled in the system.
```
CLI section has prefix-list for spt disabled group list? where is that configuration stored?

vidyac86 replied:
Added the prefix-list as well as corrected the name for spt-action

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
++       |     |  |  |        +--ro multicast-groups?   -> /oc-rpol:routing-policy/defined-sets/prefix-sets/prefix-set/name
++       |     |  |  +--rw spt-switchover
++       |     |  |  |  +--rw config
++       |     |  |  |  |  +--rw spt-action?                 enumeration
```
Hope this attr is mandatory

ravisekhar-akkineni replied:
Yes, it is. By default spt switchover is enabled (Default Val: PIM_SPT_IMMEDIATE). In order to disable spt switchover on all or a specific multicast group range, spt-action field must be set. 

When configured via REST, spt-action must be set to PIM_SPT_INFINITY before configuring spt-infinity-prefix-list where as CLI implementation takes care of the same internally

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+### 3.5.2 CLI
+
+#### 3.5.2.1 Global Configuration Commands
+
```
Please check CLI section for VRF parameter explanation, keep that consistent

vidyac86 replied:
Addressed

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+|                                  **Syntax**                                   | ip pim [vrf vrf-name] spt-threshold infinity { prefix-list prefix-list-name} |     
+|                                                                               | no ip pim [vrf vrf-name] spt-threshold infinity       |
+| **Parameters**                                                                |                                                                                       |     
+| vrf                                                                           | Applies to specific VRF if VRF input is provided       |     
```
VRF parameter explanation, keep that consistent

ravisekhar-akkineni replied:
Modified

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+|    ACL                          | Qualifiers                      |   Action                     |
+|---------------------------------|---------------------------------|------------------------------|
+| Unknown IPv4 Multicast packets must be forwarded to the CPU through a dedicated separate CoPP Queue| PacketRes = Unknown Multicast packets Destination IP=224.0.0.0/4 |   (1) CosQCpuNew -QUEUE_ID_1>  (2) GpCopyToCpu |                      |                                 |                                 |                              |
```
Is unknown multicast packets coming on any interface is taken to CPU or only on PIM enabled interfaces. I assume its only on PIM enabled interface, please confirm.

vidyac86 replied:
Yes, packets on any interface is taken to CPU. We proposed a requirement to add L3 CLASS_IDs but considering the changes in SAI, it was agreed to take up as enhancement. Not covered as part of PIM-SM changes.

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+| vrf                                                                           | Applies to specific VRF if VRF input is provided       |     
+|spt-threshold  | Shortest path tree switchover   |
+|infinity       | Disables SPT switchover
+|prefix-list   |  Provide the prefix-list or list of multicast groups on which SPT switchover has to be disabled or re-enabled.
```
" or re-enabled" - is this needed? i feel just " Provide the prefix-list or list of multicast groups on which SPT switchover has to be disabled " is fine as no cmd does not take prefix-list

ravisekhar-akkineni replied:
Modified

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+
+|ip pim spt-threshold infinity  | Command to disable or re-enable shortest path tree (SPT) swtchover for all the multicast groups or a given prefix-list.|  
+| ------------------------------------------------ | ------------------------------------------------------------ |
```
Similar to RP configuration above, please update option params syntax for prefix list as its optional

vidyac86 replied:
done

mhampasagar commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+10. Once the (S,G) routes are installed in the Hardware, the hardware starts a timer for all the installed (S,G) routes. This timer value is maintained at a lesser value than the original KAT timer and also this timer would be a global timer having the same value across all the VRFs in the hardware.  This value will be updated by FRR once it learns a first ASM route in any VRF and the value would ...
+    
+11. The ipmcFpmSyncd process will be registered for the event notification for the multicast routes in the state DB. Once the DONT_AGE/AGE status is updated in the DB, ipmcFpmSyncd will receive this event and it will update it's local DB with the age and dont_age counter values.This DB would be updated as and when there is a change in the status of the routes in the Hardware. <br><br>
```
Before these attr get updated, if pimd query for the status what will be returned? 
cases 
- where queried route not present
- the status not updated, what is the default value, is there any default value? is it same local DB or these notifications from SAI are in different local DB. if its same DB what are the default values

rathnasabapathyv replied:
@mhampasagar : I think, your question was answered in one flow diagram, which somehow was missing here. **Section : PIM KAT MFPM channel error handling.** If mroute entry is not present, when queried, that status will be reported back to FRR and then it maintains a pending list to clean/refresh it later in next cycle. That flow diagram covers other failure cases as well.

vidyac86 replied:
Added the require diagrams hence resolving this

mhampasagar reviewed (commented):
Added some comments, please check. Also please add FRR PIM design doc link for reference if we have one

rathnasabapathyv commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+![](pim_images/PIM_STATIC_RP_CONFIG_FLOW.png)
+
+## 4.2 PIM Data packet Flow
+![](2021-12-06-13-10-56.png)
```
@vidyac86 : Please check, the images for 4.2, 4.3 & 4.4 are missing. Also we had one flow diagram called "PIM KAT MFPM channel error handling". Please add that.

vidyac86 replied:
I have added them

vidyac86 committed:
Added the missing diagrams

vidyac86 committed:
Flow diagrams

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+2. Support trapping of known ipv4 multicast data to CPU through a dedicated CoPP queue.
+3. Support programming of (*,G) routes into hardware using SAI APIS.
+4. Support programming of register interface as CPU port in (*,G) and (S,G) multicast routes
+5. Support IPv4 multicast forwarding on Port based, VLAN based and LAG based routing interfaces.
```
Do we have support on sub-interfaces?

vidyac86 replied:
Yes, we support sub-interfaces as well. Updated this section.

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+6. Provide support for VRF in PIM ASM.
+7. Support the liveliness tracking of (S,G) route through HW timers and programming the HW timer value using SAI APIS.
+8. Support to stop multicast data forwarding in Kernel to avoid duplication of traffic.
+9. Support for static configuration of RP 
```
Does the 1st release have only static RP support?

vidyac86 replied:
Yes, we support only static RP

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+2. Support to configure and GET PIM-SM informations using REST and gNMI.
+3. Support for GNMI subscriber
+4. Support save and reload of PIM-SM configurations.
+5. Support to display PIM SM details in "show running-config" and also add "show running-config pim"
```
Are clear commands related to pim supported, if so, pls capture them.

vidyac86 replied:
No new clear commands are supported updated the same in requirements section

vidyac86 replied:
Updated in the requirements section as well

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+### 1.1.3 Scalability Requirements
+The following are the scalability requirements for this feature:
+1. Maximum number of PIM neighbors supported is 64.[ User can enable 64 pim interfaces in a single VRF (or) can split across 64 vrfs]   
```
From where this 64 scale number derived from?
Also in shared bcast domains/mclag env, same interface can have more than 1 neighbors on it, so is the limit on interface count or neighbor count?
Also, are we planning to return any error code when the user tries to configure PIM on more than 64 interfaces?

vidyac86 replied:
There is no explicit scaling requirements for PIM-SM. So, we have retained the same numbers, which is currently claimed for PIM-SSM.  There is no hard coded limit . User can configure more than 64 also. System will not throw error. Actually there was a restriction in earlier release where Kernel supports only 32 PIM enabled interfaces but this was increased later as part of bug fix but even then we have not claimed more than 32 PIM enabled interfaces. 
But we are yet to bench mark the numbers.

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+### 1.1.3 Scalability Requirements
+The following are the scalability requirements for this feature:
+1. Maximum number of PIM neighbors supported is 64.[ User can enable 64 pim interfaces in a single VRF (or) can split across 64 vrfs]   
+2. Maximum number of multicast route entries supported is 8K, this is inclusive of snooping and pim routes.
```
What is the max. allowed static RP count and group to RP mapping count?

vidyac86 replied:
There is no restriction on number of static RPs, since we claim only 32 interfaces can be PIM enabled we have to restrict the static RP also to that number. As we are yet to bench mark as of now we will retain it as 31.  There is no restriction in the ACL mapped to static RP

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+# 2 Functionality
+
+## 2.1 PIM-ASM Use Cases
```
Does the 1st release have support over mclag also?

vidyac86 replied:
No, we did not support on mclag in this release.

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+Once the (S,G) routes are learnt and programmed in the hardware, the traffic starts flowing through the data path. To check the liveliness of the (S,G) routes installed in the TIB, the traffic flow for the (S,G) routes in the hardware should be monitored and the status of it should be passed to the Pimd. If the traffic is not flowing for an (S,G) for the default KAT timer of 210 seconds then the  ...
+
+10. Once the (S,G) routes are installed in the Hardware, the hardware starts a timer for all the installed (S,G) routes. This timer value is maintained at a lesser value than the original KAT timer and also this timer would be a global timer having the same value across all the VRFs in the hardware.  This value will be updated by FRR once it learns a first ASM route in any VRF and the value would ...
```
What is the significance of this hw KAT timer? Could you be more specific how this plays a role in removing a route?

vidyac86 replied:
To check the liveliness of the (S,G) installed in TIB, we are not pulling the traffic to software all the time. Instead, Hardware runs a KAT timer with a less granularity than software and updates the route status to FRR through intermediate module ipmcfpmsyncd. If  H/W and S/W runs at the same time, there are chances that we might miss the route status in scaled scenario, hence the H/W KAT timer is operated at a lesser value.

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+| ip pim rp-address  |        Configure a static PIM rendezous point (RP) address for a group.                                                                     |
+| ------------------------------------------------ | ------------------------------------------------------------ |
+| **Syntax**                                                                    | ip pim [vrf vrf-name] rp-address < address > [{ prefix-list prefix-list-name}]                      |
```
Seems FRR has support for providing either prefix list or the group address range. Do we also allow the same in sonic-cli?

sonic(config)# ip pim rp 10.1.1.1
<cr>
A.B.C.D/M Group Address range to cover
prefix-list group prefix-list filter

vidyac86 replied:
No, as per ISCLI standard we are retaining only the prefix-list and not the groups in RP command.

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+|ip pim spt-threshold infinity  | Command to disable or re-enable shortest path tree (SPT) swtchover for all the multicast groups or a given prefix-list.|  
+| ------------------------------------------------ | ------------------------------------------------------------ |
+|                                  **Syntax**                                   | ip pim [vrf vrf-name] spt-threshold infinity { prefix-list prefix-list-name} |     
+|                                                                               | no ip pim [vrf vrf-name] spt-threshold infinity       |
```
Have the prefix list in no form also as an option.

vidyac86 replied:
Since we already have the configuration command to set infinity for all the groups. We are not adding the prefix-list in the no form of the command.

rajashekar-jayaraj commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+| Number of Multicast Route Entries    | 8192          |
+| Number of PIM Neighbours             | 64            |
+
+Linux kernel supports only 32 multicast routing interfaces of which one interface is reserved for PIM register interface.  So, PIM/IGMP can be enabled on only 31 routing interfaces.
```
Does it mean, we limit the number of interfaces where pim can be enabled to 31 interfaces only in sonic-cli also?

vidyac86 replied:
No, we do not limit it via CLI. There is no infrastructure available right now to do that.

vidyac86 committed:
Update SONiC_PIM_SM_HLD.md

Updated the (*,G) entry installation in OA which has loopback interface as iif

vidyac86 committed:
Update SONiC_PIM_SM_HLD.md

updated iif case for RP

vidyac86 committed:
Added the DB information of SPT

Added the DB information of SPT

vidyac86 committed:
Updated sub interface related requirements

vidyac86 committed:
Updated KAT timer details

vidyac86 committed:
Replacing the KAT flow

vidyac86 committed:
Syncd DS changes

vidyac86 committed:
Added vendor details

vidyac86 committed:
Addressed review comments

ashok-krishnegowda commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+    
+    SAI_IPMC_EVENT_DONT_AGE,
+ 
+} sai_ipmc_event_t;
```
As discussed, pls add below new event type,

/** IPMC entry deleted */
"SAI_IPMC_EVENT_DELETED"

vidyac86 committed:
syncd and datastructure naming changes

rathnasabapathyv commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1023 @@
+|  ipmc_fpm_msg_type_t |  msg_type  | message type     |
+|  uint16_t        |  msg_len   | Length of the entire message including header length |
+
+Enum mfpm_msg_type_t
```
Let's update the name for mfpm_msg_type_t. Also for mfpm_route_status_t down below. mfpm_kat_info_t few lines above. Also in some places, we refer MFPM client & MPFM channel, let's correct that as well.

vidyac86 replied:
Updated

vidyac86 committed:
Replaced mfpm with IPMC_FPM

And added details about delete notification from SAI and shaper requirements for COPP

ashok-krishnegowda commented on doc/mgmt/SONiC_PIM_SM_HLD.md:
```diff
@@ -0,0 +1,1036 @@
+
+Linux kernel supports only 32 multicast routing interfaces of which one interface is reserved for PIM register interface.  So, PIM/IGMP can be enabled on only 31 routing interfaces.
+
+# 10 Unit Test
```
Please include UT related to,
- ipmc entry aging with traffic on and off
- warmboot trigger
- config reload
- Fast boot/cold boot
- BGP docker restart
